flink-runtime/src/test/java/org/apache/flink/runtime/failure/FailureEnricherUtilsTest.java

```diff
@@ -0,0 +1,284 @@
+        assertThat(enrichers).hasSize(1);
+        // verify that the failure enricher was created and returned
+        assertThat(enrichers.iterator().next()).isInstanceOf(TestEnricher.class);
+    }
+
+    @Test
+    public void testGetValidatedEnrichers() {
+        // create two enrichers with non-overlapping keys
+        final FailureEnricher firstEnricher = new TestEnricher("key1");
+        final FailureEnricher secondEnricher = new TestEnricher("key2");
+
+        final Set<FailureEnricher> enrichers =
+                new HashSet<FailureEnricher>() {
+                    {
+                        add(firstEnricher);
+                        add(secondEnricher);
+                    }
+                };
+
+        final Collection<FailureEnricher> validatedEnrichers =
+                FailureEnricherUtils.filterInvalidEnrichers(enrichers);
+
+        // expect both enrichers to be valid
+        assertThat(validatedEnrichers).hasSize(2);
+        assertThat(validatedEnrichers).contains(firstEnricher, secondEnricher);
+    }
+
+    @Test
+    public void testValidatedEnrichersWithInvalidEntries() {
+        // create two enrichers with overlapping keys and a valid one -- must be different classes
+        final FailureEnricher validEnricher = new TestEnricher("validKey");
+        final FailureEnricher firstOverlapEnricher = new AnotherTestEnricher("key1", "key2");
+        final FailureEnricher secondOverlapEnricher = new AndAnotherTestEnricher("key2", "key3");
+
+        final Set<FailureEnricher> enrichers =
+                new HashSet<FailureEnricher>() {
+                    {
+                        add(validEnricher);
+                        add(firstOverlapEnricher);
+                        add(secondOverlapEnricher);
+                    }
+                };
+
+        final Collection<FailureEnricher> validatedEnrichers =
+                FailureEnricherUtils.filterInvalidEnrichers(enrichers);
+        // Only one enricher is valid
+        assertThat(validatedEnrichers).hasSize(1);
+    }
+
+    @Test
+    public void testLabelFutureWithValidEnricher() {
+        // validate labelFailure by enricher with correct outputKeys
+        final Throwable cause = new RuntimeException("test exception");
+        final Set<FailureEnricher> failureEnrichers = new HashSet<>();
+        final FailureEnricher validEnricher = new TestEnricher("enricherKey");
+        failureEnrichers.add(validEnricher);
+
+        final CompletableFuture<Map<String, String>> result =
+                FailureEnricherUtils.labelFailure(cause, null, failureEnrichers);
+
+        try {
+            final Map<String, String> labels = result.get();
+            assertThat(labels).hasSize(1);
+            assertThat(labels).containsKey("enricherKey");
+            assertThat(labels).containsValue("enricherKeyValue");
+        } catch (Exception e) {
+            fail("This should never happen.");
+        }
```

Comment:
you can use FlinkAssertions to simplify this (applies to multiple places)

```suggestion
        assertThatFuture(result).eventuallySucceeds().satisfies(labels -> {
            assertThat(labels).hasSize(1);
            assertThat(labels).containsKey("enricherKey");
            assertThat(labels).containsValue("enricherKeyValue");
        });
```

Reply:
Nice! Pretty useful, thanks!